# Project Overview
The goal of this project is to build a custom viral Kraken2 database that is capable of identifying plant viruses present in plant metagenomic data. When compared to a publically avaiable viral Kraken2 database. 

In order to achieve this goal, the project follows the process incorporating both Linux and R methods to generate Sankey plots and Phylogenetic trees for samples obtained from the Virginia Tech Plant Disease Clinic:
<img width="590" height="431" alt="Screenshot 2025-12-04 at 10 38 54â€¯AM" src="https://github.com/user-attachments/assets/2a364d99-1941-484e-9d98-1d9b585123ba" />
 

# Viral Reference Collection
NCBI Virus/NCBI Datasets API, go over the general process required to collect the sequences. Complete and partially complete genomes since those results look the best and I need to build the trees anyways. (Only use the three samples I showed during the research update lab meeting). Focus on the scripts, needs to be similar to the class project done earlier in the semester, build the Sankey plots with RStudio.

# Kraken2 Database Building
Best to mention contaminated sequences (mention the process of using BLAST to remove any sequence that contains plant DNA).

# Sankey plot comparison between custom database and publically available

# Extracting abundant virus, assembly, and IQtree

# Bootstrap tree generation with RStudio ggtree

# All used code needs to have comments explaining what is going on.
