# Project Overview
The goal of this project is to build a custom viral Kraken2 database that is capable of identifying plant viruses present in plant metagenomic data. When compared to a publically avaiable viral Kraken2 database. 

In order to achieve this goal, the project follows the process incorporating both Linux and R methods to generate Sankey plots and Phylogenetic trees for samples obtained from the Virginia Tech Plant Disease Clinic:
<img width="590" height="431" alt="Screenshot 2025-12-04 at 10 38 54 AM" src="https://github.com/user-attachments/assets/2a364d99-1941-484e-9d98-1d9b585123ba" />
 
# Kraken2 Custom Database Building
The Kraken2 custom database used for this pipeline was generated by collecting all complete and partially complete viral sequences from known plant-infecting viral families from NCBI using the NCBI datasets Conda package. CheckV was run on both the complete sequences and partially complete sequences to determine the completeness score of the partially complete sequences. Any partially complete sequence that did not fall within a specified range of the average sequence length of the complete sequences belonging to each plant viral family were removed from the database. Any partially complete sequence that had a completeness percentage below 30% was also removed from the database (30% was used based on CheckV stating any sequence below 30% complete is considered low-quality). To remove any presence of plant DNA within the viral reference sequences, a BLAST was run on both the complete and partially complete sequences against a BLAST DB of only plant genomes. Any viral sequence that has a positive hit against the plant DB was removed from the Kraken2 custom database. After sequence filtering, the resulting sequences were built into a Kraken 2 database using the Kraken 2 Conda package (specifically the k2 wrapper commands as well as kraken2-build):

```
k2 download-taxonomy --db DB_NAME

for file in PATH/TO/REFERENCE/SEQUENCES/*.fna; do
    echo "Adding $file..."
    kraken2-build --add-to-library "$file" --db DB_NAME
done

k2 build --db DB_NAME
k2 clean --db DB_NAME
```

# Conda Environment Setup
Create the conda environment:
```
  module load Miniconda3
  conda env create -f environment.yml -n indiv_prod_mg
```

# Linux Script
<details>
  <summary>Click to expand code</summary>
  
```

```
</details>

# Sankey Plot Comparison
Sankey plots were generated from the Kraken2 report file from the Kraken2 classification command ran in the Linux script (this process is done manually outside of the linux script). The sankey plots were created using the Pavian Shiny App Webinterface ([https://fbreitwieser.shinyapps.io/pavian/)](https://fbreitwieser.shinyapps.io/pavian/). The sankey plots from the Kraken2 custom database and the Kraken2 viral RefSeq databases are shown below for sample 24_557:

**Viral RefSeq Database**
<img width="722" height="381" alt="Screenshot 2025-12-16 at 12 24 43 PM" src="https://github.com/user-attachments/assets/f1a5c9e3-cbd9-4822-a18a-0e07a58b0639" />

**Custom Database**
<img width="722" height="381" alt="Screenshot 2025-12-16 at 12 25 11 PM" src="https://github.com/user-attachments/assets/835fec54-7f21-4539-99c0-697583cb8e40" />

# R Script
The R script below goes through the process of generating a phylogenetic tree visualization of an IQtree consensus tree with 1000 tree replicates. The tree is generated with ggtree and dyprl primarily. The script takes an input tree file, input metadata on the sample of interest (includes geographic location, host organism, virus name, etc.), and a taxonomic lineage file for all viral species used in the Kraken 2 custom database. The metadata and taxonomic lineage file were generated with the NCBI Datasets Conda package (Not shown in the scripts). The resulting tree for sample 24_557 is shown below:
<details>
  <summary>Click to expand code</summary>
  
```

```
</details>

# Notes to myself
One large issue I find with my coding and script writing is to more frequently use comments in my code as well as staying better organized with my folders and the locations of scripts/inputs/outputs. I did not realize how much I struggled with this before taking this course.
The other issue I find with my coding is writing scripts in a sample to sample basis rather than a loop to at the start. If I began writing scripts with loops in mind, I would save myself a lot of time converting a non-loop-based script into a loop-based script.
